src/core/common/settings.cpp

```diff
@@ -0,0 +1,584 @@
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the OpenThread platform abstraction for non-volatile storage of settings.
+ *
+ */
+
+#include <stdlib.h>
+#include <stddef.h>
+#include <string.h>
+#include <assert.h>
+
+#include <openthread-types.h>
+#include <openthread-core-config.h>
+
+#include <common/code_utils.hpp>
+#include <platform/settings.h>
+#include <platform/flash.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+enum
+{
+    BLOCK_ADD_FLAG = 0x1,
+    BLOCK_DELETE_FLAG = 0x2,
+};
+
+enum
+{
+    MAX_STAGE_ADD_NUM = 3,
+    MAX_STAGE_DELETE_NUM = 3,
+    MAX_STAGE_DATA_LEN = 32,
+};
+
+struct settingsBlock
+{
+    uint16_t key;
+    uint8_t index;
+    uint8_t flag;
+    uint16_t length;
+    uint16_t reserved;
+};
+
+struct stageAddSettingsBlock
+{
+    struct settingsBlock block;
+    uint8_t data[MAX_STAGE_DATA_LEN];
+};
+
+struct stageDeleteSettingsBlock
+{
+    uint16_t key;
+    int index;
+};
+
+
+static uint32_t sAddress;
+static uint32_t sSettingsBlockFlag = 0xbe5cc5ef;
+static bool sCommitLock = false;
+
+static uint16_t sStageActionSeq;
+static struct stageAddSettingsBlock sStageAddSettingsBlock[MAX_STAGE_ADD_NUM];
+static uint8_t sStageAddSettingsNum;
+static uint16_t sStageAddSettingsBufLength;
+static struct stageDeleteSettingsBlock sStageDeleteSettingsBlock[MAX_STAGE_DELETE_NUM];
+static uint8_t sStageDeleteSettingsNum;
+
+void reorderSettingsBlock(void)
+{
+    uint8_t ramBuffer[OPENTHREAD_CONFIG_SETTINGS_PAGE_SIZE];
```

Comment:
Another way is to always use  2* N flash pages. N 's default is 1, but can be larger like 2, 3....